### PR TITLE
Optimize boxart image load

### PR DESCRIPTION
**Pull request by networkfusion**
## Description

Improves boxart image loading by using directory file paths.
It also adds matching by full game ID's (for compatibility), but notes in the readme that it is slow.

## Motivation and Context

Loading boxart was slow in certain circumstances.

## How Has This Been Tested?

Locally on an SC64.

## Types of changes

- [x] Improvement (non-breaking change that adds a new feature)
- [ ] Bug fix (fixes an issue)
- [ ] Breaking change (breaking change)
- [ ] Documentation Improvement
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Signed-off-by: GITHUB_USER <GITHUB_USER_EMAIL>

## Summary by CodeRabbit

- **New Features**
	- Introduced a new image type enumeration for boxart components, enhancing image management.
	- Updated boxart initialization to allow for specific image views, improving flexibility in image retrieval.
	- Enhanced instructions for setting up boxart images with clearer directory structures and simplified naming conventions.

- **Bug Fixes**
	- Improved the logic for loading boxart images, enabling better fallback options in case of missing files.

### CODE REVIEW

**Comment by networkfusion** [comment]
@coderabbitai review

**Comment by networkfusion** [comment]
@coderabbitai review

**Comment by networkfusion** [comment]
@Suprapote are you the author of the boxart packages? if so, they could be vastly improved here...

**Comment by Suprapote** [comment]
i'm the creator of the European and 64DD boxarts. What type of improvements do you want?

**Comment by networkfusion** [comment]
> i'm the creator of the European and 64DD boxarts. What type of improvements do you want?

Use 4 letter ID's, even better, use the new path schema (as explained in the readme of this PR)

**Comment by networkfusion** [comment]
Thinking about conversion:

```
param (
    [switch]$Convert4CharFiles
)

# Get all .png files in the current directory
$files = Get-ChildItem -Filter "*.png"

foreach ($file in $files) {
    # Extract the filename without extension
    $filename = [System.IO.Path]::GetFileNameWithoutExtension($file.Name)
    
    # Check the length of the filename
    if ($filename.Length -eq 4) {
        if ($Convert4CharFiles) {
            # Split the filename into individual characters
            $chars = $filename.ToCharArray()
            
            # Create the new directory path
            $newPath = Join-Path -Path (Join-Path -Path (Join-Path -Path $chars[0] $chars[1]) $chars[2]) $chars[3]
        } else {
            continue
        }
    } elseif ($filename.Length -eq 3) {
        # Split the filename into individual characters
        $chars = $filename.ToCharArray()
        
        # Create the new directory path
        $newPath = Join-Path -Path (Join-Path -Path $chars[0] $chars[1]) $chars[2]
    } elseif ($filename.Length -eq 2) {
        if ($filename -eq "ZL") {
            # Create the new directory path with prefix 'C'
            $newPathC = Join-Path -Path (Join-Path -Path "C" $filename[0]) $filename[1]
            # Create the new directory path with prefix 'N'
            $newPathN = Join-Path -Path (Join-Path -Path "N" $filename[0]) $filename[1]
            
            # Create the new directories if they don't exist
            if (-not (Test-Path -Path $newPathC)) {
                New-Item -ItemType Directory -Path $newPathC -Force
            }
            if (-not (Test-Path -Path $newPathN)) {
                New-Item -ItemType Directory -Path $newPathN -Force
            }
            
            # Define the new file paths
            $newFilePathC = Join-Path -Path $newPathC -ChildPath "boxart_front.png"
            $newFilePathN = Join-Path -Path $newPathN -ChildPath "boxart_front.png"
            
            # Move the file to both locations with the new name
            Copy-Item -Path $file.FullName -Destination $newFilePathC
            Move-Item -Path $file.FullName -Destination $newFilePathN
            continue
        } else {
            # Split the filename into individual characters
            $chars = $filename.ToCharArray()
            
            # Create the new directory path with prefix 'N'
            $newPath = Join-Path -Path (Join-Path -Path "N" $chars[0]) $chars[1]
        }
    } else {
        # Skip filenames that are not 2, 3, or 4 characters long
        continue
    }
    
    # Create the new directory if it doesn't exist
    if (-not (Test-Path -Path $newPath)) {
        New-Item -ItemType Directory -Path $newPath -Force
    }
    
    # Define the new file path
    $newFilePath = Join-Path -Path $newPath -ChildPath "boxart_front.png"
    
    # Move the file to the new location with the new name
    Move-Item -Path $file.FullName -Destination $newFilePath
}
```

```
.\YourScriptName.ps1 -Convert4CharFiles
```

File: README.md [hunk 1]
```diff
@@ -49,22 +49,43 @@ An open source menu for N64 flashcarts.
...
+These must be `PNG` files that use the following dimensions:
+* Standard N64 GamePak boxart sprites: 158x112
+* Japanese N64 GamePak boxart sprites: 112x158
+* 64DD boxart sprites: 129x112
+
+They will be loaded by directories using each character of the full 4 character Game Code (as identified in the menus ROM information).
+i.e. for GoldenEye NTSC USA (NGEE), this would be `sd:/menu/boxart/N/G/E/E/boxart_front.png`.
+i.e. for GoldenEye PAL (NGEP), this would be `sd:/menu/boxart/N/G/E/P/boxart_front.png`.
+
+To improve compatibility between regions, you may exclued the region ID (last matched directory) for GamePaks to match with 3 letter IDs instead:
```
Comment (networkfusion) [inline README.md:67]:
```suggestion
To improve compatibility between regions (as a fallback), you may exclued the region ID (last matched directory) for GamePaks to match with 3 letter IDs instead:
```

File: src/menu/components/boxart.c [hunk 2]
```diff
@@ -29,21 +29,89 @@ component_boxart_t *component_boxart_init (const char *storage_prefix, char *gam
...
-
-    // TODO: This is bad, we should only check for 3 letter codes
-    sprintf(file_name, "%.2s.png", game_code + 1);
-    path_push(path, file_name);
-    if (png_decoder_start(path_get(path), BOXART_WIDTH_MAX, BOXART_HEIGHT_MAX, png_decoder_callback, b) == PNG_OK) {
-        path_free(path);
-        return b;
+
+    if (directory_exists(path_get(path))) {
+        if (current_image_view == IMAGE_GAMEPAK_FRONT) {
```
Comment (networkfusion) [inline src/menu/components/boxart.c:40]: perhaps use a switch statement instead?

**Pull request merged**